### PR TITLE
Add per-atom weights

### PR DESCRIPTION
**Pull request by luigibonati**
- [x] Added per-atom weights
- [x] Charge CV in ASE calculator

### CODE REVIEW

File: mace/calculators/mace.py [hunk 1]
```diff
@@ -234,23 +333,37 @@ def calculate(self, atoms=None, properties=None, system_changes=all_changes):
...
         for i, model in enumerate(self.models):
             batch = batch_base.clone()
-            if self.model_type in ["AtomicChargesMACE", "EnergyChargesMACE"]:
-                out = model(
-                    batch.to_dict(),
-                    compute_stress=compute_stress,
-                )
-            else:
-                out = model(batch.to_dict(), compute_stress=compute_stress)
+            model_out = model(
```
Comment (luigibonati) [inline mace/calculators/mace.py:341]: here you are first computing energy and charges with`model.forward` and then the derivatives (both forces and of the charge cv) with `self._get_outputs`, right?

i find it a little bit confusing that some of the outputs come from `model_out` and some from `out`. for instance the energy is passed through `self._get_outputs` while the charges and the node_energies are not. maybe it is better to pass everything that has been computed from forward to self._get_outputs and then access them from the same dict?

File: mace/calculators/mace.py [hunk 2]
```diff
@@ -310,12 +423,24 @@ def calculate(self, atoms=None, properties=None, system_changes=all_changes):
             self.results["charges"] = (
                 torch.mean(ret_tensors["charges"], dim=0).cpu().numpy()
             )
+            if (self.charge_cv_expr is not None):
+                self.results["charge_cv"] = (
+                    torch.mean(ret_tensors["charge_cvs"], dim=0).cpu().numpy()
+                )
+                self.results["charge_cv_gradients"] = (
+                    torch.mean(ret_tensors["charge_cv_gradients"], dim=0).cpu().numpy()
```
Comment (luigibonati) [inline mace/calculators/mace.py:433]: why are you taking the mean in the charge_cv and charge_cv_gradients expressions?

File: mace/data/utils.py [hunk 3]
```diff
@@ -120,6 +123,14 @@ def config_from_atoms(
     stress = atoms.info.get(stress_key, None)  # eV / Ang
     virials = atoms.info.get(virials_key, None)
     dipole = atoms.info.get(dipole_key, None)  # Debye
+    group = atoms.arrays.get(group_key, None)
```
Comment (luigibonati) [inline mace/data/utils.py:126]: can we use the `tags` property which is already present by default in `ase.atoms` instead of creating a new one? 
https://wiki.fysik.dtu.dk/ase/ase/atoms.html#ase.Atoms.get_tags

File: mace/data/atomic_data.py [hunk 4]
```diff
@@ -201,6 +210,30 @@ def from_config(
...
+            raise RuntimeError('atom_group_weights and element_group_weights should not be specified at the same time!')
+        elif (atom_group_weights is not None and config.group is None):
+            raise RuntimeError('atom_group_weights is specified but can not find group definitions in configuration!')
+        elif (atom_group_weights is not None):
+            group_list = config.group.tolist()
+            peratom_weights = torch.ones(len(config.positions), dtype=torch.get_default_dtype())
+            for i in set(config.group):
+                if (i not in atom_group_weights.keys()):
+                    raise RuntimeError('can not find weight for group {:d}!'.format(i))
+                weight_avg = atom_group_weights[i] / group_list.count(i)
```
Comment (luigibonati) [inline mace/data/atomic_data.py]: regarding the normalization of the weights, if we have a dictionary like: {'Fe':1, 'N':1}

the actual weights will be 1/num(Fe) and 1/num(N), and then later normalized to sum to 1, right? so to give the same weights to each chemical species we should know the (average) number of atoms

File: mace/calculators/mace.py (hunk 1 above)
Comment (jintuzhang) [reply mace/calculators/mace.py:341]: > here you are first computing energy and charges with model.forward and then the derivatives (both forces and of the charge cv) with self._get_outputs, right?

That's right.

> maybe it is better to pass everything that has been computed from forward to self._get_outputs and then access them from the same dict?

Sure. This has been resolved in the latest commit.

File: mace/calculators/mace.py (hunk 2 above)
Comment (jintuzhang) [reply mace/calculators/mace.py:433]: I don't know. The original ASE calculator calculates the mean values for other qualities (energy, forces) when multiple models were given. I just did this to match the existed style.

File: mace/data/utils.py (hunk 3 above)
Comment (jintuzhang) [reply mace/data/utils.py:126]: We can do this, but I think using the generic `atoms.arrays.get` method gives us greater flexibilities. When you want to keep the `tag` data field in your dataset, using another data array should be useful.

Besides, setting data array in ASE is simple enough. You could just do somethings like `atoms.set_array('charges', [0] * 10 + [1] * 20, int, ())`

File: mace/data/atomic_data.py (hunk 4 above)
Comment (jintuzhang) [reply mace/data/atomic_data.py]: > the actual weights will be 1/num(Fe) and 1/num(N), and then later normalized to sum to 1, right? so to give the same weights to each chemical species we should know the (average) number of atoms

Yes exactly, we could do this in some other manner, but I feel that there are trade-offs in the ease of use.

File: mace/calculators/mace.py (hunk 2 above)
Comment (luigibonati) [reply mace/calculators/mace.py:433]: I see it! so the mean is across multiple models, thanks!
